_LowLevelCode/cpp/cpp_communicator/MPI_wrapper.cpp

```diff
@@ -23,20 +23,34 @@ int MPI_wrapper::init(const InitParamHolder &init_param) {
     int* argc(nullptr);
     char*** argv(nullptr);
     int err(-1);
-    // initiate the assynchroneous messages queue.
+    // initiate the asynchronous messages queue.
     this->async_queue_max_len_ = init_param.async_queue_length;
     this->asyncMessList.clear();
     //
     if (init_param.is_tested) {
-        // set up test values and return without initializeing the framework
+        // set up test values and return without initializing the framework
         this->isTested = true;
         this->labIndex = (int)init_param.debug_frmwk_param[0];
         this->numLabs = (int)init_param.debug_frmwk_param[1];
         this->SyncMessHolder.resize(this->numLabs);
         this->InterruptHolder.resize(this->numLabs);
+        this->node_names.resize(this->numLabs);
+        if (this->labIndex == 0) {
+            char node_name[100];
+            for (int i = 0; i < this->numLabs; i++) {
+                sprintf(node_name, "Node%d\n", i + 1);
+                this->node_names[i].assign(node_name, strlen(node_name + 1));
+            }
+        }
         return 0;
     }
     try {
+        int is_initialized;
+        MPI_Initialized(&is_initialized);
+        if (is_initialized) {
+            throw_error("MPI_MEX_COMMUNICATOR:runtime_error",
+                "MPI framework is initialized before MPI init was invoked");
+        }
         err = MPI_Init(argc, argv);
     }
     catch (...) {}
```

Comment:
Global catches are very bad. Could throw for a number of reasons, not just the one you've highlighted. At the very least set `err` to not be `MPI_SUCCESS`

Reply:
check for initialized probably should not be there. Moved out. Otherwise have no idea what `MPI_Init` could throw. 
Its not declared `nothrow` so may be C is not throwing. Can not be sure about all MPI implementations